src/base/memory_monitor_sr.cpp

```diff
@@ -312,9 +352,52 @@ namespace cubperf
       }
   }
 
-  int mmon_module::aggregate_stats (const MMON_MODULE_INFO &info)
+  int mmon_module::aggregate_stats (MMON_MODULE_INFO &info, bool get_summary) const
   {
     int error = NO_ERROR;
+
+    strncpy (info.name, m_module_name.c_str (), m_module_name.size ());
+
+    if (get_summary)
+      {
+	info.stat.cur_stat = m_stat.cur_stat.load ();
+	return error;
+      }
+    else
+      {
+	info.stat.init_stat = m_stat.init_stat.load ();
+	info.stat.cur_stat = m_stat.cur_stat.load ();
+	info.stat.peak_stat = m_stat.peak_stat.load ();
+	info.stat.expand_resize_count = m_stat.expand_resize_count.load ();
+
+
+	info.num_comp = m_component.size ();
+	if (info.num_comp != 0)
+	  {
+	    info.comp_info = (MMON_COMP_INFO *)malloc (sizeof (MMON_COMP_INFO) * info.num_comp);
+
+	    if (info.comp_info == NULL)
+	      {
+		er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, sizeof (MMON_COMP_INFO) * info.num_comp);
+		error =  ER_OUT_OF_VIRTUAL_MEMORY;
+	      }
+
+	    for (uint32_t i = 0; i < info.num_comp; i++)
+	      {
+		error = m_component[i]->get_stat (info.comp_info[i]);
```

Comment:
A positive `error` will be overwritten by following one. For example
I = 0 : error
I = 1 : no_error
-> It has to realize when I == 0.

I think we should check the error every get_stat() and clean up resources we have allocated at that point.

Reply:
If it's too picky to track allocation and do delicate according to error cases, I think you can use unique_ptr<A[]>. or you can just use std::vector.

Reply:
I'm good with yours. I just mentioned another way (seems easier to me :) ).